zebra-state/src/service.rs

```diff
@@ -252,6 +252,87 @@ impl StateService {
     pub fn utxo(&self, outpoint: &transparent::OutPoint) -> Option<transparent::Output> {
         self.mem.utxo(outpoint).or_else(|| self.sled.utxo(outpoint))
     }
+
+    /// Return an iterator over the relevant chain of the block identified by
+    /// `hash_or_height`.
+    #[allow(dead_code)]
+    pub fn chain(&self, hash_or_height: HashOrHeight) -> Iter<'_> {
+        Iter {
+            service: self,
+            state: State::NonFinalized(hash_or_height),
+        }
+    }
+}
+
+struct Iter<'a> {
+    service: &'a StateService,
+    state: State,
+}
+
+enum State {
+    NonFinalized(HashOrHeight),
+    Finalized(block::Height),
+    Finished,
+}
+
+impl<'a> Iter<'a> {
+    fn next_non_finalized_block(&mut self) -> Option<Arc<Block>> {
+        let Iter { service, state } = self;
+
+        let hash_or_height = match state {
+            State::NonFinalized(hash_or_height) => hash_or_height,
+            State::Finalized(_) | State::Finished => unreachable!(),
+        };
+
+        if let Some(block) = service.mem.block(*hash_or_height) {
+            let hash = block.header.previous_block_hash;
+            self.state = State::NonFinalized(hash.into());
+            Some(block)
+        } else {
+            None
+        }
+    }
+
+    fn next_finalized_block(&mut self) -> Option<Arc<Block>> {
+        let Iter { service, state } = self;
+
+        let hash_or_height = match *state {
+            State::Finalized(height) => height.into(),
+            State::NonFinalized(hash_or_height) => hash_or_height,
```

Comment:
It's a bit confusing for the finalized state function to access the `NonFinalized` enum variant.

Instead, let's have an additional enum variant for newly initialised iterators, before the first call to `next`.

Reply:
Oh, right, we need to access that enum variant when transitioning from non-finalized to finalized state.

I don't think this is a problem then.

Reply:
I started with that but that wouldn't apply here. The reason we access `NonFinalizedState` here isn't because it's newly setup, it's because the `hash` used in the last `next_non_finalized_state` is not a hash of non-finalized state, and so we need to use that same hash to lookup the height in the finalized state.

By accessing the HashOrHeight(hash) from the previous state at the start of `next_finalized_block` we make it so we only have to lookup the height for the hash once, and then we iterate by height which should be more sympathetic to the database design than continuing to iterate by previous hash.

We could change it so that if the `next_non_finalized_block` fails to find the block, it first attempts to resolve the height of the hash via the finalized state, but this would be doing work that wont be needed until the next iteration, which may never occur, so my preference would be to keep the current design.